server/ipqueue.go

```diff
@@ -185,62 +215,94 @@ func (q *ipQueue[T]) resetAndReturnToPool(elts *[]T) {
 func (q *ipQueue[T]) popOne() (T, bool) {
 	q.Lock()
+	if q.elts == nil {
+		q.Unlock()
+		var empty T
+		return empty, false
+	}
 	l := len(q.elts) - q.pos
-	if l < 1 {
+	if l == 0 {
 		q.Unlock()
 		var empty T
 		return empty, false
 	}
 	e := q.elts[q.pos]
-	q.pos++
-	if q.calc != nil {
-		q.sz -= q.calc(e)
-	}
 	l--
 	if l > 0 {
+		q.pos++
+		if q.calc != nil {
+			q.sz -= q.calc(e)
+		}
 		// We need to re-signal
 		select {
 		case q.ch <- struct{}{}:
 		default:
 		}
 	} else {
 		// We have just emptied the queue, so we can recycle now.
-		q.resetAndReturnToPool(&q.elts)
-		q.elts, q.pos = nil, 0
+		if cap(q.elts) > q.mrs {
+			q.elts = nil
+		} else {
+			q.elts = q.elts[:0]
+		}
+		q.pos, q.sz = 0, 0
 	}
 	q.Unlock()
 	return e, true
 }
 
-// After a pop(), the slice can be recycled for the next push() when
+// After a `pop()`, the slice can be recycled for the next `push()` when
 // a first element is added to the queue.
-// This will also decrement the "in progress" count with the length
-// of the slice.
-// Reason we use pointer to slice instead of slice is explained
-// here: https://staticcheck.io/docs/checks#SA6002
-func (q *ipQueue[T]) recycle(elts *[]T) {
-	// If invoked with a nil list, nothing to do.
-	if elts == nil || *elts == nil {
+// This will also decrement the "in progress" count/size with the given
+// `ql` and `qsz` which were returned by `pop()` (and possibly updated
+// if calling `processed()`).
+func (q *ipQueue[T]) recycle(elts []T, ql int64, qsz uint64) {
+	if q == nil || elts == nil {
 		return
 	}
-	// Update the in progress count.
-	if len(*elts) > 0 {
-		if atomic.AddInt64(&q.inprogress, int64(-(len(*elts)))) < 0 {
-			atomic.StoreInt64(&q.inprogress, 0)
-		}
+	if qsz > 0 {
+		atomic.AddUint64(&q.inprogressSz, ^(qsz - 1))
+	}
+	if ql > 0 {
+		atomic.AddInt64(&q.inprogress, -ql)
 	}
 	// We also don't want to recycle huge slices, so check against the max.
 	// q.mrs is normally immutable but can be changed, in a safe way, in some tests.
-	if cap(*elts) > q.mrs {
+	c := cap(elts)
+	if c > q.mrs {
 		return
 	}
-	q.resetAndReturnToPool(elts)
+	q.psMu.Lock()
+	if len(q.pool) < ipQueueMaxPoolSize && q.tcps+c <= q.mrs {
+		elts = elts[:0]
+		q.pool = append(q.pool, elts)
+		q.tcps += c
+	}
+	q.psMu.Unlock()
+}
+
+// When a queue has limits specified, `push()` will take into account the number
+// and size of "in progress" elements, that is, the ones removed from a `pop()` call.
+// When processing those elements, the caller should call `q.processed()` on
+// individual elements to decrement the number and size of "in progress" elements.
+func (q *ipQueue[T]) processed(e T, ql *int64, qsz *uint64) {
+	if calc := q.calc; calc != nil {
+		if sz := calc(e); sz > 0 {
+			atomic.AddUint64(&q.inprogressSz, ^(sz - 1))
+			*qsz -= sz
+		}
+	}
+	atomic.AddInt64(&q.inprogress, -1)
+	*ql--
 }
 
 // Returns the current length of the queue.
 func (q *ipQueue[T]) len() int {
 	q.Lock()
 	defer q.Unlock()
+	if q.elts == nil {
```

Comment:
This was when I started to have `q.elts` as a `*[]T`, but not necessary anymore, `return len(q.elts) - q.pos` would work even if `q.elts` is `nil`, so I could remove that.